src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpUtilities.cs

```diff
@@ -81,72 +80,87 @@ private static ulong GetMaskAsLong(ReadOnlySpan<byte> bytes)
         return BinaryPrimitives.ReadUInt64LittleEndian(bytes);
     }
 
-    // The same as GetAsciiStringNonNullCharacters but throws BadRequest
+    // The same as GetAsciiString but throws BadRequest for null character
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    public static unsafe string GetHeaderName(this ReadOnlySpan<byte> span)
+    public static string GetHeaderName(this ReadOnlySpan<byte> span)
     {
         if (span.IsEmpty)
         {
             return string.Empty;
         }
 
-        fixed (byte* source = &MemoryMarshal.GetReference(span))
+        var str = string.Create(span.Length, span, static (destination, source) =>
         {
-            return string.Create(span.Length, new IntPtr(source), s_getHeaderName);
-        }
-    }
-
-    private static unsafe void GetHeaderName(Span<char> buffer, IntPtr state)
-    {
-        fixed (char* output = &MemoryMarshal.GetReference(buffer))
-        {
-            // This version of AsciiUtilities returns null if there are any null (0 byte) characters
-            // in the string
-            if (!StringUtilities.TryGetAsciiString((byte*)state.ToPointer(), output, buffer.Length))
+            if (Ascii.ToUtf16(source, destination, out var written) != OperationStatus.Done
+                || destination.Contains('\0'))
```

Comment:
```suggestion
            if (source.Contains((byte)0) ||
                Ascii.ToUtf16(source, destination, out var written) != OperationStatus.Done)
```
`source` is half the length of `destination` so it can be searched faster; since do both on sucess might as well check source first as it will be the faster of the two ops